cmd/image-autobumper/imagebumper/imagebumper.go

```diff
@@ -0,0 +1,281 @@
+}
+
+// FindLatestTag returns the latest valid tag for the given image.
+func (cli *Client) FindLatestTag(imageHost, imageName, currentTag string) (string, error) {
+	k := imageHost + "/" + imageName + ":" + currentTag
+	if result, ok := cli.tagCache[k]; ok {
+		return result, nil
+	}
+
+	currentTagParts := tagRegexp.FindStringSubmatch(currentTag)
+	if currentTagParts == nil {
+		return "", fmt.Errorf("couldn't figure out the current tag in %q", currentTag)
+	}
+	if currentTagParts[tagVersionPart] == "latest" {
+		return currentTag, nil
+	}
+
+	imageList, err := cli.getManifest(imageHost, imageName)
+	if err != nil {
+		return "", err
+	}
+
+	latestTag, err := pickBestTag(currentTagParts, imageList)
+	if err != nil {
+		return "", err
+	}
+
+	cli.tagCache[k] = latestTag
+
+	return latestTag, nil
+}
+
+func (cli *Client) TagExists(imageHost, imageName, currentTag string) (bool, error) {
+	imageList, err := cli.getManifest(imageHost, imageName)
+	if err != nil {
+		return false, err
+	}
+
+	for _, v := range imageList {
+		for _, tag := range v.Tags {
+			if tag == currentTag {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func pickBestTag(currentTagParts []string, manifest manifest) (string, error) {
+	// The approach is to find the most recently created image that has the same suffix as the
+	// current tag. However, if we find one called "latest" (with appropriate suffix), we assume
+	// that's the latest regardless of when it was created.
+	var latestTime int64
+	latestTag := ""
+	for _, v := range manifest {
+		bestVariant := ""
+		override := false
+		for _, t := range v.Tags {
+			parts := tagRegexp.FindStringSubmatch(t)
+			if parts == nil {
+				continue
+			}
+			if parts[tagExtraPart] != currentTagParts[tagExtraPart] {
+				continue
+			}
+			if parts[tagVersionPart] == "latest" {
+				override = true
+				continue
+			}
+			if bestVariant == "" || len(t) < len(bestVariant) {
+				bestVariant = t
+			}
+		}
+		if bestVariant == "" {
+			continue
+		}
+		t, err := strconv.ParseInt(v.TimeCreatedMs, 10, 64)
+		if err != nil {
+			return "", fmt.Errorf("couldn't parse timestamp %q: %w", v.TimeCreatedMs, err)
+		}
+		if override || t > latestTime {
+			latestTime = t
+			latestTag = bestVariant
+			if override {
+				break
+			}
+		}
+	}
+
+	if latestTag == "" {
+		return "", fmt.Errorf("failed to find a good tag")
+	}
+
+	return latestTag, nil
+}
+
+// AddToCache keeps track of changed tags
+func (cli *Client) AddToCache(image, newTag string) {
+	cli.tagCache[image] = newTag
+}
+
+func updateAllTags(tagPicker func(host, image, tag string) (string, error), content []byte, imageFilter *regexp.Regexp) []byte {
+	indexes := imageRegexp.FindAllSubmatchIndex(content, -1)
+	// Not finding any images is not an error.
+	if indexes == nil {
+		return content
+	}
+
+	newContent := make([]byte, 0, len(content))
+	lastIndex := 0
+	for _, m := range indexes {
+		newContent = append(newContent, content[lastIndex:m[imageTagPart*2]]...)
```

Comment:
Cryptic. 
And the references in `m` are constant, so we can define (for readability) `imageHostPartStart`, `imageHostStartEnd` etc.